### PR TITLE
Add history export feature

### PR DESCRIPTION
**Pull request by molikuner**
This would allow to export the history as CSV file. Currently it works, but the code doesn't look good and only exports the content. I would like to also export the time and the type.
Also it allows to export binary data to CSV, does this even make any sense? Suggestions are welcome.

**TODO:**
- [x] Escape content if needed
- [x] Reorganize usage of the db to be accessed by a repo/usecase ?!
- [x] Add some real res strings, don't use hard-coded stuff...

### CODE REVIEW

**Comment by markusfisch** [comment]
Just to give you a quick feedback on this; I think you can't do it this way 😞 See, a `Cursor` _needs_ to be closed with `cursor.close()` to release its resources. Not closing it is a resource leak. In Java, lint throws a warning about this and also various other static analyzers (e.g. https://github.com/facebook/infer). Kotlin's still quite new and not on par with Java's tooling unfortunately 😬

**Comment by molikuner** [comment]
I know about the `.close()` function. That's the reason why I use `cursor.use()` ([docs](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html)). This safely closes the cursor even if an exception is thrown. 
Or did I forgot it anywhere?

**Comment by markusfisch** [comment]
No, you're right, of course! I just saw you deleting those `cursor.close()` lines and jumped right onto it without double checking 🙈 `cursor.use {}` is fine of course!

**Comment by molikuner** [comment]
As we already are at the problem of closing something. Are you sure that we doesn't need to close the db itself? Its closable (implements SQLiteClosable).

**Comment by markusfisch** [comment]
Yes, I am, but it's a good question!

To understand why, it's good to know that Android's [SQLiteDatabase](https://android.googlesource.com/platform/frameworks/base.git/+/43a17654cf4bfe7f1ec22bd8b7b32daccdf27c09/core/java/android/database/sqlite/SQLiteDatabase.java) is just an object keeping database connections of the native implementation of SQLite. SQLite is designed in such a way, that an application crash will not corrupt any data (as long as you don't do anything funny that is, see https://www.sqlite.org/howtocorrupt.html). A lost connection is not a problem.

So, closing a `SQLiteDatabase` just means releasing resources. If there is more than just one database, this makes sense. But if there's just one database *and* we're using it often, it's much simpler to keep the `SQLiteDatabase` around than to try and manage opening/closing manually. If the resources blocked by `SQLiteDatabase` are too much, the system may as well kill the whole app.

And that's my main point: Android can (and will) terminate any app if it's short on resources. Manually managing database connections would just add complexity which makes things more fragile. SQLite, on the other hand, is able to deal with abnormal program terminations. So I think it's best to keep things simple and just keep the `SQLiteDatabase` instance available as long as the process exists.

**Comment by molikuner** [comment]
Okay, thanks for the explanation. I just thought we could close it in `onDestroy` of the application as it's opened in `onCreate`.

Feel free to review the changes and merge afterwards.

**Comment by markusfisch** [comment]
No, we can't because [Application](https://developer.android.com/reference/android/app/Application) has no method `onDestroy` 😉 There's just [onTerminate](https://developer.android.com/reference/android/app/Application#onTerminate()) which is never called on a production device. So, yes, there's no callback/method/event that is triggered when an app (or rather its process) is being killed by the system 🤷‍♂

But since system resources are automatically managed by the system (in contrast to a desktop system), an app shouldn't really need that. Getting killed lies in the nature of a mobile app 😉 

Will review and merge tomorrow.

**Comment by molikuner** [comment]
Well, I didn't know about the lacking of an `onDestroy` function :smile:. I thought there would be one, just like for Activity, Fragment, etc.

File: app/src/main/kotlin/de/markusfisch/android/binaryeye/fragment/HistoryFragment.kt [hunk 1]
```diff
@@ -168,6 +183,63 @@ class HistoryFragment : Fragment() {
...
+					).show()
+				}
+			}
+		}
+	}
+
+	private fun List<DatabaseRepository.Scan>.toCSV(delimiter: String, allowBinary: Boolean): ByteArray {
+		val result = mutableListOf<Byte>()
+
+		val header = "DATE${delimiter}TYPE${delimiter}CONTENT${if (allowBinary) "${delimiter}BINARY_CONTENT" else ""}\n".toByteArray()
```
Comment (molikuner) [inline app/src/main/kotlin/de/markusfisch/android/binaryeye/fragment/HistoryFragment.kt]: I'm not really happy with this method of creating a csv, but I don't have any good idea on how to improve it. There are 2 locations in the code where the order is defined. As you can see from the last [commit](0617695) it's bad for maintainability, as a code change requires an other change in an other place without compiler error. Any ideas on improvement?
I could refactor, to just have one function, but that's not as reusable and creates a big unclean/unreadable function.

Comment (molikuner) [reply app/src/main/kotlin/de/markusfisch/android/binaryeye/fragment/HistoryFragment.kt]: Found a way: ffe92eb

**Comment by markusfisch** [comment]
Well, I'm not really a fan of the `DatabaseRepository` pattern because, in my eyes, it just adds another layer of complexity. A layer, that makes things a little bit easier to handle, admittedly. But it also uses up more memory, at least when we are storing all the rows from a cursor in a List. With a small set of records, it won't be a problem. But the bigger the set gets, the more memory pressure there is, for no other reason than that it's easier to deal with the data.

A cursor, on the other hand, has a low memory profile because it doesn't need to have the whole list in memory. Which is a bit slower, but scales much better.

That said, I think the actual implementation of `DatabaseRepository` is beautiful 👍 Kotlin is just great in making complex things look simple. What doesn't make complex things necessarily better 😉 

Now, I've merge all the changes and we will see if `DatabaseRepository` is up to the task. It would've been better, if this change was its own commit (so we could revert it should it become necessary) but I guess it's a bit too late for that 😉

**Comment by molikuner** [comment]
Yeah, I'm aware of the higher impact on memory caused by that change. I tried to implement it as sequence to possibly save some memory space, but that didn't worked, because the Cursor was closed before actual use :confused:.
Maybe we could update `kotlinx-coroutines` and use `Flows` for that.

**Pull request merged**